shared/helper/adm_preset_definitions_helper.cpp

```diff
@@ -0,0 +1,608 @@
+#include "adm_preset_definitions_helper.h"
+#include <adm/common_definitions.hpp>
+#include <adm/parse.hpp>
+#include <algorithm>
+#include <string>
+#include <string_view>
+#include <sstream>
+#include <vector>
+#include <helper/container_helpers.hpp>
+#include <helper/supplementary_definitions.hpp>
+#include <helper/cartesianspeakerlayouts.h>
+
+namespace {
+	std::vector<std::string_view> splitString(const std::string& input, char delimiter, size_t maxElements) {
+		std::vector<std::string_view> result;
+		std::string_view str(input);
+		size_t start = 0;
+
+		for (size_t i = 0; i < maxElements - 1; ++i) {
+			size_t end = str.find(delimiter, start);
+			if (end == std::string_view::npos) {
+				break;
+			}
+			result.push_back(str.substr(start, end - start));
+			start = end + 1;
+		}
+
+		// Add the remaining text, even if it contains more delimiters
+		result.push_back(str.substr(start));
+
+		return result;
+	}
+
+	template <typename Prop, typename Elm>
+	bool matchingProp(const Elm& elmA, const Elm& elmB) {
+		bool hasProp = elmA.template has<Prop>();
+		if (hasProp != elmB.template has<Prop>())
+			return false;
+		if (hasProp) {
+			auto valA = elmA.template get<Prop>();
+			auto valB = elmB.template get<Prop>();
+			if (valA != valB) return false;
+		}
+		return true;
+	}
```

Comment:
TL/DR here's an alternative implementation of the comparison bit, which you can adopt or ignore.

I noticed doing it that there's no has() check on the ScreenEdgeLock on the original implementation, so if nothing else that probably needs adding in as I think it's an optional property, so the subsequent get() calls would fail if a speaker position of either flavour does not have ScreenEdgeLock set on a custom preset.

I mentioned I got nerd-sniped by this yesterday, my refactor is below.
It separates out the various stages
  - pulling out the first block format from an audiochannelformat / checking it's the only one
  - checking whether presence of a property (element.has()) on a pair of parent elements match
  - checking whether properties of a certain type are equal
  - combining the presence and equality checks for a parent pair and a property
  - combining checks for multiple properties on a parent pair (with short circuit &&, equivalent to the early return that was there before)
 
This hopefully makes each stage more simple than it was previously, and it extends easily to other elements by filling in the missing special case `equals()` functions.

Currently dumped into an anonymous namespace (roughly equivalent to making all the helpers 'static'), but it probably makes sense to complete & stick it in libadm at some point, which is why I popped it in an 'equality' namespace (it would eventually be adm::equality). I've previously thought about just implementing operator== everywhere, but I'm not sure if desired behaviour is always the same (should defaulted properties and those that are explicitly set to the same value compare equal?), and using a different name/namespace makes it easier for a user to opt-in or replace the functionality themselves than is the case if we define operator==() in adm::

```c++
namespace {
namespace equality {
using namespace adm;

enum class CompareHasResult {
  Neither, Both, Mismatch };

template<typename Property, typename Parent>
CompareHasResult compareHas(Parent const& lhs, Parent const& rhs) {
  bool hasProp = lhs.template has<Property>();
  if (hasProp != rhs.template has<Property>())
    return CompareHasResult::Mismatch;
  if (hasProp) {
    return CompareHasResult::Both;
  }
  return CompareHasResult::Neither;
}

template<typename T>
bool equals(T const& lhs, T const& rhs) {
  return std::equal_to<T>{}(lhs, rhs);
}

template<typename Property, typename Parent>
bool propertyMatch(Parent const& lhs, Parent const& rhs) {
  switch(compareHas<Property>(lhs, rhs)) {
  case CompareHasResult::Mismatch:
    return false;
  case CompareHasResult::Neither:
    return true;
  case CompareHasResult::Both:
    return equals(lhs.template get<Property>(), rhs.template get<Property>());
  }
}

template<typename... Properties, typename Parent>
bool propertiesMatch(Parent const& lhs, Parent const& rhs) {
  return (propertyMatch<Properties, Parent>(lhs, rhs) && ...);
}

template<>
bool equals(Gain const& lhs, Gain const& rhs) {
  return lhs.get() == rhs.get();
}

template<>
bool equals(SpeakerLabels const& lhs, SpeakerLabels const& rhs) {
  return std::equal(lhs.begin(), lhs.end(), rhs.begin(), rhs.end());
}

template<>
bool equals(HeadphoneVirtualise const& lhs, HeadphoneVirtualise const& rhs) {
  return propertiesMatch<Bypass, DirectToReverberantRatio>(lhs, rhs);
}

template<>
bool equals(ScreenEdgeLock const& lhs, ScreenEdgeLock const& rhs) {
  return propertiesMatch<HorizontalEdge, VerticalEdge>(lhs, rhs);
}

template<>
bool equals(SphericalSpeakerPosition const& lhs, SphericalSpeakerPosition const& rhs) {
  return propertiesMatch<Azimuth, AzimuthMin, AzimuthMax, Elevation, ElevationMin, ElevationMax, Distance, DistanceMin, DistanceMax, ScreenEdgeLock>(lhs, rhs);
}

template<>
bool equals(CartesianSpeakerPosition const& lhs, CartesianSpeakerPosition const& rhs) {
  return propertiesMatch<X, XMin, XMax, Y, YMin, YMax, Z, ZMin, ZMax, ScreenEdgeLock>(lhs, rhs);
}
}

std::optional<adm::AudioBlockFormatDirectSpeakers const> static_ds_block(adm::AudioChannelFormat const& cf) {
  if(cf.get<adm::TypeDescriptor>() == adm::TypeDefinition::DIRECT_SPEAKERS) {
    auto blocks = cf.getElements<adm::AudioBlockFormatDirectSpeakers>();
    if(blocks.size() == 1) {
      return blocks.front();
    }
  }
  return {};
}
}

bool AdmPresetDefinitionsHelper::isEquivalentByProperties(std::shared_ptr<const adm::AudioChannelFormat> cfA, std::shared_ptr<const adm::AudioChannelFormat> cfB)
{
  using namespace adm;
  if(equality::compareHas<TypeDescriptor>(*cfA, *cfB) !=
     equality::CompareHasResult::Both) {
    return false;
  }
  // TODO: only support matching up DS AudioChannelFormats for now
  auto bfA = static_ds_block(*cfA);
  auto bfB = static_ds_block(*cfB);
  if(bfA && bfB) {
    return equality::propertiesMatch<SpeakerLabels, Importance, HeadLocked, Gain, HeadphoneVirtualise, SphericalSpeakerPosition, CartesianSpeakerPosition>(*bfA, *bfB);
  }
  return false;
}
```

Reply:
I think I prefer your second solution. I'm not sure the first makes the code any more readable. With the second, I can stick all that in another header to tidy up this file (including isEquivalentByProperties, come to think of it as that's a general ADM helper too)